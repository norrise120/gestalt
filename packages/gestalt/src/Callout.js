// @flow strict
import * as React from 'react';
import PropTypes, { string } from 'prop-types';
import Box from './Box.js';
import Icon from './Icon.js';
import IconButton from './IconButton.js';
import Link from './Link.js';
import Text from './Text.js';
import typography from './Typography.css';

type Props = {|
  description: string,
    iconAccessibilityLabel: string,
      primaryLink ?: {|
        href: string,
          label: string,
  |},
onDismiss: () => void,
  secondaryLink ?: {|
    href: string,
      label: string,
  |},
style: 'error' | 'info' | 'warning',
  title ?: string,
|};

const CALLOUT_TYPE_ATTRIBUTES = {
  info: {
    icon: 'check-circle',
    color: 'blue',
    backgroundColor: '#EBF4FE',
  },
  warning: {
    icon: 'workflow-status-warning',
    color: 'orange',
    backgroundColor: '#FDF5EC',
  },
  error: {
    icon: 'workflow-status-problem',
    color: 'red',
    backgroundColor: '#FDEBEE',
  },
};

export default function Callout({
  description,
  iconAccessibilityLabel,
  onDismiss,
  primaryLink,
  secondaryLink,
  style,
  title,
}: Props): React.Node {
  return (
    <Box
      alignItems="center"
      display="flex"
      position="relative"
      rounding={4}
      dangerouslySetInlineStyle={{
        __style: { backgroundColor: CALLOUT_TYPE_ATTRIBUTES[style].backgroundColor }
      }}
      padding={8}
    >
      <Box>
        <Icon icon={CALLOUT_TYPE_ATTRIBUTES[style].icon} color={CALLOUT_TYPE_ATTRIBUTES[style].color} accessibilityLabel="" size={24} />
      </Box>
      <Box flex="grow" paddingX={3}>
        {title && (
          <Box marginBottom={2}>
            <Text weight="bold">{title}</Text>
          </Box>
        )}
        <Text>{description}</Text>
      </Box>
      {secondaryLink && (
        <Box padding={4} flex="none">
          <Link href={secondaryLink.href}>
            <Text weight="bold">{secondaryLink.label}</Text>
          </Link>
        </Box>
      )}
      {primaryLink && (
        <Box rounding="pill" color="white" padding={4} flex="none">
          <Link href={primaryLink.href}>
            <Text weight="bold">{primaryLink.label}</Text>
          </Link>
        </Box>
      )}
      <Box position="absolute" top right padding={1}>
        <IconButton icon="cancel" iconColor="darkGray" size="sm" accessibilityLabel="" onClick={onDismiss} />
      </Box>
    </Box>
  )
}

Callout.propTypes = {
  description: PropTypes.string.isRequired,
  iconAccessibilityLabel: PropTypes.string.isRequired,
  primaryLink: PropTypes.exact({
    href: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
  }),
  onDismiss: PropTypes.func,
  secondaryLink: PropTypes.exact({
    href: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
  }),
  style: PropTypes.oneOf(['error', 'info', 'warning']).isRequired,
  title: PropTypes.string,
}