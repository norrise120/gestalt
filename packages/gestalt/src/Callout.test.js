// @flow strict
import React from 'react';
import { create } from 'react-test-renderer';
import Callout from './Callout.js';

describe('<Callout />', () => {
  test('Error Callout', () => {
    const tree = create(
      <Callout
        description="Insert a clever error callout message here"
        iconAccessibilityLabel="error icon"
        style="error"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Warning Callout', () => {
    const tree = create(
      <Callout
        description="Insert a clever warning callout message here"
        iconAccessibilityLabel="warning icon"
        style="warning"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Info Callout', () => {
    const tree = create(
      <Callout
        description="Insert a clever info callout message here"
        iconAccessibilityLabel="info icon"
        style="info"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('description + title', () => {
    const tree = create(
      <Callout
        description="Insert a clever info callout message here"
        iconAccessibilityLabel="info icon"
        style="info"
        title="A Title"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('description + title + primaryLink', () => {
    const tree = create(
      <Callout
        description="Insert a clever info callout message here"
        iconAccessibilityLabel="info icon"
        primaryLink={{ href: 'pinterest.com', label: 'Visit Pinterest' }}
        style="info"
        title="A Title"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('description + title + primaryLink + secondaryLink', () => {
    const tree = create(
      <Callout
        description="Insert a clever info callout message here"
        iconAccessibilityLabel="info icon"
        primaryLink={{ href: 'pinterest.com', label: 'Visit Pinterest' }}
        secondaryLink={{ href: 'pinterest.com/help', label: 'Learn more' }}
        style="info"
        title="A Title"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('description + title + primaryLink + dismissButton', () => {
    const tree = create(
      <Callout
        description="Insert a clever info callout message here"
        iconAccessibilityLabel="info icon"
        primaryLink={{ href: 'pinterest.com', label: 'Visit Pinterest' }}
        dismissButton={{
          accessibilityLabel: 'Dismiss banner',
          onDismiss: () => {},
        }}
        style="info"
        title="A Title"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
