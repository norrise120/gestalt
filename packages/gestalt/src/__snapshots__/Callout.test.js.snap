// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Callout /> Error Callout 1`] = `
<div
  className="box itemsCenter paddingX8 paddingY8 relative rounding4 xsDisplayFlex"
  style={
    Object {
      "backgroundColor": "#FDEBEE",
    }
  }
>
  <div
    className="box"
  >
    <svg
      aria-hidden={true}
      aria-label=""
      className="icon red iconBlock"
      height={24}
      role="img"
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        d="test-file-stub"
      />
    </svg>
  </div>
  <div
    className="box flexGrow paddingX3"
  >
    <div
      className="Text fontSize3 darkGray alignLeft breakWord fontWeightNormal"
    >
      Insert a clever error callout message here
    </div>
  </div>
  <div
    className="absolute box paddingX1 paddingY1 right0 top0"
  >
    <button
      aria-label="Dismiss banner"
      className="button tapTransition enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <div
        className="pog transparent"
        style={
          Object {
            "height": 32,
            "width": 32,
          }
        }
      >
        <svg
          aria-hidden={true}
          aria-label=""
          className="icon darkGray iconBlock"
          height={16}
          role="img"
          viewBox="0 0 24 24"
          width={16}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`<Callout /> Info Callout 1`] = `
<div
  className="box itemsCenter paddingX8 paddingY8 relative rounding4 xsDisplayFlex"
  style={
    Object {
      "backgroundColor": "#EBF4FE",
    }
  }
>
  <div
    className="box"
  >
    <svg
      aria-hidden={true}
      aria-label=""
      className="icon blue iconBlock"
      height={24}
      role="img"
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        d="test-file-stub"
      />
    </svg>
  </div>
  <div
    className="box flexGrow paddingX3"
  >
    <div
      className="Text fontSize3 darkGray alignLeft breakWord fontWeightNormal"
    >
      Insert a clever info callout message here
    </div>
  </div>
  <div
    className="absolute box paddingX1 paddingY1 right0 top0"
  >
    <button
      aria-label="Dismiss banner"
      className="button tapTransition enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <div
        className="pog transparent"
        style={
          Object {
            "height": 32,
            "width": 32,
          }
        }
      >
        <svg
          aria-hidden={true}
          aria-label=""
          className="icon darkGray iconBlock"
          height={16}
          role="img"
          viewBox="0 0 24 24"
          width={16}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`<Callout /> Warning Callout 1`] = `
<div
  className="box itemsCenter paddingX8 paddingY8 relative rounding4 xsDisplayFlex"
  style={
    Object {
      "backgroundColor": "#FDF5EC",
    }
  }
>
  <div
    className="box"
  >
    <svg
      aria-hidden={true}
      aria-label=""
      className="icon orange iconBlock"
      height={24}
      role="img"
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        d="test-file-stub"
      />
    </svg>
  </div>
  <div
    className="box flexGrow paddingX3"
  >
    <div
      className="Text fontSize3 darkGray alignLeft breakWord fontWeightNormal"
    >
      Insert a clever warning callout message here
    </div>
  </div>
  <div
    className="absolute box paddingX1 paddingY1 right0 top0"
  >
    <button
      aria-label="Dismiss banner"
      className="button tapTransition enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <div
        className="pog transparent"
        style={
          Object {
            "height": 32,
            "width": 32,
          }
        }
      >
        <svg
          aria-hidden={true}
          aria-label=""
          className="icon darkGray iconBlock"
          height={16}
          role="img"
          viewBox="0 0 24 24"
          width={16}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`<Callout /> description + title + primaryLink + secondaryLink 1`] = `
<div
  className="box itemsCenter paddingX8 paddingY8 relative rounding4 xsDisplayFlex"
  style={
    Object {
      "backgroundColor": "#EBF4FE",
    }
  }
>
  <div
    className="box"
  >
    <svg
      aria-hidden={true}
      aria-label=""
      className="icon blue iconBlock"
      height={24}
      role="img"
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        d="test-file-stub"
      />
    </svg>
  </div>
  <div
    className="box flexGrow paddingX3"
  >
    <div
      className="box marginBottom2"
    >
      <div
        className="Text fontSize3 darkGray alignLeft breakWord fontWeightBold"
      >
        A Title
      </div>
    </div>
    <div
      className="Text fontSize3 darkGray alignLeft breakWord fontWeightNormal"
    >
      Insert a clever info callout message here
    </div>
  </div>
  <div
    className="box flexNone paddingX4 paddingY4"
  >
    <a
      className="link tapTransition touchable block rounding0 hoverUnderline"
      href="pinterest.com/help"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      rel=""
      target={null}
    >
      <div
        className="Text fontSize3 darkGray alignLeft breakWord fontWeightBold"
      >
        Learn more
      </div>
    </a>
  </div>
  <div
    className="box flexNone paddingX4 paddingY4 pill whiteBg"
  >
    <a
      className="link tapTransition touchable block rounding0 hoverUnderline"
      href="pinterest.com"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      rel=""
      target={null}
    >
      <div
        className="Text fontSize3 darkGray alignLeft breakWord fontWeightBold"
      >
        Visit Pinterest
      </div>
    </a>
  </div>
  <div
    className="absolute box paddingX1 paddingY1 right0 top0"
  >
    <button
      aria-label="Dismiss banner"
      className="button tapTransition enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <div
        className="pog transparent"
        style={
          Object {
            "height": 32,
            "width": 32,
          }
        }
      >
        <svg
          aria-hidden={true}
          aria-label=""
          className="icon darkGray iconBlock"
          height={16}
          role="img"
          viewBox="0 0 24 24"
          width={16}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`<Callout /> description + title + primaryLink 1`] = `
<div
  className="box itemsCenter paddingX8 paddingY8 relative rounding4 xsDisplayFlex"
  style={
    Object {
      "backgroundColor": "#EBF4FE",
    }
  }
>
  <div
    className="box"
  >
    <svg
      aria-hidden={true}
      aria-label=""
      className="icon blue iconBlock"
      height={24}
      role="img"
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        d="test-file-stub"
      />
    </svg>
  </div>
  <div
    className="box flexGrow paddingX3"
  >
    <div
      className="box marginBottom2"
    >
      <div
        className="Text fontSize3 darkGray alignLeft breakWord fontWeightBold"
      >
        A Title
      </div>
    </div>
    <div
      className="Text fontSize3 darkGray alignLeft breakWord fontWeightNormal"
    >
      Insert a clever info callout message here
    </div>
  </div>
  <div
    className="box flexNone paddingX4 paddingY4 pill whiteBg"
  >
    <a
      className="link tapTransition touchable block rounding0 hoverUnderline"
      href="pinterest.com"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      rel=""
      target={null}
    >
      <div
        className="Text fontSize3 darkGray alignLeft breakWord fontWeightBold"
      >
        Visit Pinterest
      </div>
    </a>
  </div>
  <div
    className="absolute box paddingX1 paddingY1 right0 top0"
  >
    <button
      aria-label="Dismiss banner"
      className="button tapTransition enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <div
        className="pog transparent"
        style={
          Object {
            "height": 32,
            "width": 32,
          }
        }
      >
        <svg
          aria-hidden={true}
          aria-label=""
          className="icon darkGray iconBlock"
          height={16}
          role="img"
          viewBox="0 0 24 24"
          width={16}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;

exports[`<Callout /> description + title 1`] = `
<div
  className="box itemsCenter paddingX8 paddingY8 relative rounding4 xsDisplayFlex"
  style={
    Object {
      "backgroundColor": "#EBF4FE",
    }
  }
>
  <div
    className="box"
  >
    <svg
      aria-hidden={true}
      aria-label=""
      className="icon blue iconBlock"
      height={24}
      role="img"
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        d="test-file-stub"
      />
    </svg>
  </div>
  <div
    className="box flexGrow paddingX3"
  >
    <div
      className="box marginBottom2"
    >
      <div
        className="Text fontSize3 darkGray alignLeft breakWord fontWeightBold"
      >
        A Title
      </div>
    </div>
    <div
      className="Text fontSize3 darkGray alignLeft breakWord fontWeightNormal"
    >
      Insert a clever info callout message here
    </div>
  </div>
  <div
    className="absolute box paddingX1 paddingY1 right0 top0"
  >
    <button
      aria-label="Dismiss banner"
      className="button tapTransition enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <div
        className="pog transparent"
        style={
          Object {
            "height": 32,
            "width": 32,
          }
        }
      >
        <svg
          aria-hidden={true}
          aria-label=""
          className="icon darkGray iconBlock"
          height={16}
          role="img"
          viewBox="0 0 24 24"
          width={16}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;
